R/S3.R

```diff
@@ -124,38 +124,94 @@ is_S3_class <- function(x) {
 }
 
 # -------------------------------------------------------------------------
-# Define a few base examples
+# Pull out validation functions so hit by code coverage
 
-S3_factor <- new_S3_class("factor",
-  function(.data = integer(), levels = character()) {
+validate_factor <- function(self) {
+  c(
+    if (typeof(self) != "integer")
+      "Underlying data must be an <integer>",
+    if (!is.character(attr(self, "levels")))
+      "attr(, 'levels') must be a <character>"
+  )
+}
+
+validate_date <- function(self) {
+  if (!is.numeric(self)) {
+    "Underlying data must be numeric"
+  }
+}
+
+validate_POSIXct <- function(self) {
+  if (!is.numeric(self)) {
+    return("Underlying data must be numeric")
+  }
+
+  tz <- attr(self, "tz")
+  if (!is.character(tz) || length(tz) != 1) {
+    return("attr(, 'tz') must be a single string")
+  }
+}
+
+validate_data.frame <- function(self) {
+  if (!is.list(self)) {
+    return("Underlying data must be a <list>")
+  }
+
+  rn <- attr(self, "row.names")
+  if (!is.character(rn) && !is.integer(rn)) {
+    return("attr(, 'row.names') must be a <character> or <integer>")
+  }
```

Comment:
Hmm, doesn't this materialize compact row names? We have to use something special in dplyr to avoid this, i.e.

```r
> dplyr:::dplyr_reconstruct.data.frame
function (data, template) 
{
    attrs <- attributes(template)
    attrs$names <- names(data)
    attrs$row.names <- .row_names_info(data, type = 0L)
    attributes(data) <- attrs
    data
}
```

Reply:
Oh yeah, good catch. It turns out that `structure()` won't let you set non-character/integer row names, so I can just use `type = 2` here to check that the length is correct.